### DISCUSSION TITLE
SvelteKit Library in turborepo

### DISCUSSION THREAD

**Discussion by beeb** [sveltejs/kit#7559] [Q&A]
I am trying to setup a monorepo with turborepo and pnpm where two apps built on SvelteKit will use a shared UI components library.

I tried setting up the UI library with create-svelte. In my `lib/index.ts` I try to import my only `Button.svelte` component, which is in the same folder, as `import Button from './Button.svelte'` but even this gives me a TS error: 

`Cannot find module './Button.svelte' or its corresponding type declarations.ts(2307)` although it builds fine.

Then, when trying to add this to my two apps as such:

```js
// package.json
{
        // ...
	"devDependencies": {
                // ...
		"ui": "workspace:*"
	},
	"type": "module"
}
```

I get a similar error on importing: `Cannot find module 'ui' or its corresponding type declarations.ts(2307)`

When running the dev server, I get:

```
[commonjs--resolver] Failed to resolve entry for package "ui". The package may have incorrect main/module/exports specified in its package.json.
```

Any help with setting up a SvelteKit library would be highly appreciated!

Thanks

**Comment by WillsterJohnson** [comment]
This happens because `workspace:*` looks for the workspace project `package.json`, whereas the distributable library's `package.json` is auto-generated by sveltekit.

Basically, assuming your ui package is located at `ROOT/packages/ui`, depending on `"ui": "workspace:*"` will try to import the module via `ROOT/packages/ui/package.json`.
In reality though, the package script has placed the dist code at `ROOT/packages/ui/package`, and the correct place to import the module from is via `ROOT/packages/ui/package/package.json`.

As for workarounds, I'm not sure whats best. You could add a workspace to your monorepo at `packages/*/package/`, but this would lead to technically duplicate packages, depending on your setup this might be okay, or it might not.

A possibility I haven't looked into (I haven't had the need/chance to) is reexporting `package.json`.
Try updating your `ROOT/packages/ui/package.json` to include the following;
```json
{
  "exports": {
    "./package.json": "./package/package.json"
  }
}
```
This might successfully trick node into using the `./package` directory without having to write any scripts to restructure the distribution code.

**Reply by beeb** [reply]
Hey there! Thanks for the answer. I tried the second solution (re-exporting the packaged json), it seems to work when building, but fails to refresh when the file is updated in dev mode. I'll have a deeper look when I'm less drunk :D

**Comment by The-Noah** [comment]
I'm facing the same issue, have you managed to solve it?

**Reply by beeb** [reply]
No sorry, still no luck, I've reconsidered using a monorepo at all because I'm having issues with other stuff like go serverless functions on vercel etc.

**Comment by WillsterJohnson** [comment]
I'm currently working on a project management CLI which aims to make the sveltekit monorepo experience seamless. It's still being designed but I've got a lot of the groundwork finalised. I'll share the link to the project once it's ready for use.

**Reply by adonaicandido** [reply]
any updates on this?

**Reply by WillsterJohnson** [reply]
I've decided to drop the project as `svelte-package` now does the vast majority of the work my tool was planned to do.

Instead of generating a new `package.json`, `svelte-package` now validates the existing one, meaning you can just add building your lib as a dependent step for running your app server and you need no magic or hacks to make things work.

FWIW, [here's the repo with the incomplete tool](https://github.com/willsterjohnson/sk-mono). I said "vast majority" above, the rest of the tool's functionality was going to be an opinionated project structure which used symlinks and other such magic. I've since turned away from that because it's a pain to get it to play nice.

I probably won't revisit the project, or at least if I do it will have very different priorities.

**Reply by WillsterJohnson** [reply]
Y'know since I posted the above it's been on my mind a fair bit.

I'm going to revisit the idea, try to cut out the stuff that's not strictly necessary and add in some things which would actually fill a gap in the workflow.
No promises that anything will come of this, but if I haven't given an update here by, let's say July 15th, please feel free to @ me and ask.

**Reply by shaictal** [reply]
@WillsterJohnson il take you up on this! :)
